app/src/main/java/org/greenstand/android/TreeTracker/managers/Locations.kt

```diff
@@ -0,0 +1,329 @@
+            it.hasAccuracy() && it.accuracy < ValueHelper.MIN_ACCURACY_DEFAULT_SETTING
+        } ?: false
+    }
+
+    private fun hasLocationPermissions(): Boolean {
+
+        val fineLocationPermission = ContextCompat.checkSelfPermission(
+            context,
+            android.Manifest.permission.ACCESS_FINE_LOCATION
+        )
+        val coarseLocationPermission = ContextCompat.checkSelfPermission(
+            context,
+            android.Manifest.permission.ACCESS_COARSE_LOCATION
+        )
+        return (fineLocationPermission == PackageManager.PERMISSION_GRANTED ||
+                coarseLocationPermission == PackageManager.PERMISSION_GRANTED)
+    }
+}
+
+enum class Accuracy {
+    GOOD,
+    BAD,
+    NONE
+}
+
+fun Location?.accuracyStatus(): Accuracy {
+    if (this == null || !hasAccuracy()) {
+        return Accuracy.NONE
+    }
+    return if (accuracy < ValueHelper.MIN_ACCURACY_DEFAULT_SETTING) {
+        Accuracy.GOOD
+    } else {
+        Accuracy.BAD
+    }
+}
+
+class LocationDataCapturer(
+    private val userManager: UserManager,
+    private val locationUpdateManager: LocationUpdateManager,
+    private val treeTrackerDAO: TreeTrackerDAO
+) {
+    var convergenceWithinRange: Boolean = false
+
+    private val gson = GsonBuilder().serializeNulls().create()
+    private var lastNLocations: Deque<Location> = LinkedList<Location>()
+    var generatedTreeUuid: UUID? = null
+        private set
+    var convergence: Convergence? = null
+
+    private val locationObserver: Observer<Location?> = Observer {
+        it?.apply {
+
+            MainScope().launch(Dispatchers.IO) {
+                val locationData =
+                    LocationData(
+                        userManager.planterCheckinId,
+                        latitude,
+                        longitude,
+                        accuracy,
+                        generatedTreeUuid?.toString() ?: null,
+                        System.currentTimeMillis()
+                    )
+                Timber.d("Convergence: Generated Location Data value $locationData")
+                val jsonValue = gson.toJson(locationData)
+                Timber.d("Convergence: Inserting a new location data $jsonValue")
+                treeTrackerDAO.insertLocationData(LocationDataEntity(jsonValue))
+            }
+
+            if (isInTreeCaptureMode() && !(convergenceWithinRange)) {
+
+                var evictedLocation: Location? = if (lastNLocations.size >= CONVERGENCE_DATA_SIZE)
+                    lastNLocations.pollFirst() else null
+                lastNLocations.add(it)
+
+                if (lastNLocations.size >= CONVERGENCE_DATA_SIZE) {
+                    if (convergence == null) {
+                        convergence = Convergence(lastNLocations.toList())
+                        convergence!!.computeConvergence()
+                    } else {
+                        convergence!!.computeSlidingWindowConvergence(evictedLocation!!, it)
```

Comment:
Evicted location won't be null  because `locationsDeque` will be of full size required for convergence for the initial computation. Subsequent computations would pop the head of the queue which becomes the evicted location as per line 173.